### PR TITLE
Cleanup/rework class requirements

### PR DESCRIPTION
**Pull request by davidneto92**
To better handle filtering and finding spells based on their respective classes and levels, the `class_requirements` attribute has been broken into two models.
 - `PathfinderClass` - stores the class name.
 - `SpellLevel` - stores the level and references to a spell and class. This has a many through relationship.

The tests have been corrected, but I am not sure how to filter by class and level.

### CODE REVIEW

File: app/models/pathfinder_class.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class PathfinderClass < ApplicationRecord
+  has_many :spell_levels
+  has_many :spells, through: :spell_levels
+
+  CLASSES = %w(
```
Comment (brian-penguin) [inline app/models/pathfinder_class.rb:5]: Should probably add a validation that the class is included in this list

File: app/models/spell.rb [hunk 2]
```diff
@@ -11,44 +14,17 @@ class Spell < ApplicationRecord
...
-
   validates :name, presence: true
   validates :school, inclusion: { in: SCHOOLS }
   validates :casting_time, presence: true
   validates :source, presence: true
   validates :dismissible, inclusion: { in: [true, false] }
   validates :description, presence: true
   validates :spell_requirements, length: { is: 4 }
-  validates :class_requirements, length: { is: 24 }
+  # validates :class_requirements, length: { is: 24 }
```
Comment (brian-penguin) [inline app/models/spell.rb]: Keep or remove don't leave comments

File: app/models/spell.rb [hunk 3]
```diff
@@ -11,44 +14,17 @@ class Spell < ApplicationRecord
...
 
   validate :validate_spell_requirements
-  validate :validate_class_requirements
+  # validate :validate_class_requirements
```
Comment (brian-penguin) [inline app/models/spell.rb]: same here

File: app/models/spell.rb [hunk 4]
```diff
@@ -77,13 +53,13 @@ def list_spell_requirements
       end
     end
 
-    if display.include?("f")
-      display = display.split("").sort.reverse!
-    else
-      display = display.split("")
-    end
+    
+    display = if display.include?("f")
```
Comment (brian-penguin) [inline app/models/spell.rb]: Isn't this the change we merged the other day?

File: db/seeds.rb [hunk 5]
```diff
@@ -8,70 +10,52 @@ def boolean_check(value)
...
 
-def build_spell_requirements(spell)
-  return {
-    verbal:   boolean_check(spell[:verbal]),
-    somatic:  boolean_check(spell[:somatic]),
-    material: boolean_check(spell[:material]),
-    focus:    boolean_check(spell[:focus])
-  }
+def build_class_requirements(spell_data)
+  PathfinderClass.all.each do |klass|
```
Comment (brian-penguin) [inline db/seeds.rb]: I would call this `pathfinder_class`

`klass` is generally reserved for meta programming involving class names

File: spec/services/spell_finder_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  let!(:params) { { class_name: 'paladin' } }
+  let!(:spell_finder) { SpellFinder.new(params) }
+
+  it 'index can be filtered to show all spells of a class' do
+    expect(spell_finder.perform).to match_array([paladin_spell])
+  end
+
+  pending 'index can be filtered to show spells of a specific level and class'
```
Comment (brian-penguin) [inline spec/services/spell_finder_spec.rb]: we don't need a pending spec with nothing in it

File: spec/factories/pathfinder_class.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :pathfinder_class do
+    sequence(:name) { |n| "Pathfinder Class name 0#{n}" }
```
Comment (brian-penguin) [inline spec/factories/pathfinder_class.rb]: This will need to be updated with the validations

**Review by brian-penguin — CHANGES_REQUESTED** [review]
Some small changes, adding violations and cleaning up specs then good.

File: app/models/spell.rb (hunk 4 above)
Comment (davidneto92) [reply app/models/spell.rb]: Yeah, I ended up having to copy it so Rubocop would allow it.

File: db/seeds.rb (hunk 5 above)
Comment (davidneto92) [reply db/seeds.rb]: Understood, makes sense to preserve that convention.

File: app/models/pathfinder_class.rb (hunk 1 above)
Comment (davidneto92) [reply app/models/pathfinder_class.rb:5]: Validations added for uniqueness and inclusion in list of classes.

File: spec/factories/pathfinder_class.rb (hunk 7 above)
Comment (davidneto92) [reply spec/factories/pathfinder_class.rb]: Corrected with `.sample`

**Comment by davidneto92** [comment]
Comments noted and corrections made.

**Comment by davidneto92** [comment]
Comments noted and rework complete. Merging.

**Pull request merged**